src/main/java/at/ac/tuwien/kr/alpha/grounder/IndexedInstanceStorage.java

```diff
@@ -168,6 +169,41 @@ public void removeInstance(Instance instance) {
 		return matchingInstances == null ? new ArrayList<>() : matchingInstances;
 	}
 
+
+	private int getFirstGroundTermPosition(Atom atom) {
+		for (int i = 0; i < atom.getTerms().size(); i++) {
+			Term testTerm = atom.getTerms().get(i);
+			if (testTerm.isGround()) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	public Collection<Instance> getInstancesFromPartiallyGroundAtom(Atom substitute) {
+		// For selection of the instances, find ground term on which to select.
+		int firstGroundTermPosition = getFirstGroundTermPosition(substitute);
+		// Select matching instances, select all if no ground term was found.
+		if (firstGroundTermPosition != -1) {
+			Term firstGroundTerm = substitute.getTerms().get(firstGroundTermPosition);
+			return getInstancesMatchingAtPosition(firstGroundTerm, firstGroundTermPosition);
+		} else {
+			return new ArrayList<>(getAllInstances());
+		}
+	}
+
+	public Collection<Instance> getInstancesMatching(Atom matchAtom) {
+		ArrayList<Instance> matchingInstances = new ArrayList<>();
```

Comment:
It is a common convention to have the more general/abstract collection type on the left, e.g.:
```java
List<T> xs = new ArrayList<T>();
```
instead of
```java
ArrayList<T> xs = new ArrayList<T>();
```
and
```java
Set<T> xs = new HashSet<T>();
```
instead of
```java
HashSet<T> xs = new HashSet<T>();
```

The main reason being that this way the code that then actually deals with `xs` must comply with the more general behaviour, and cannot depend on the concrete implementation being used. The implementation to be used is only relevant at initialization time. Of course there are exceptions when one really needs the concrete implementing type throughout, but in most cases you just want a list or a set, and that suffices. The special cases where one depends on the implementation of a collection class is a gain something that should be justified separately in my opinion.